gaputil.1
.\" -*- nroff -*-
.\"
.\" Copyright (C) 2015 Bradley Worley <geekysuavo@gmail.com>.
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to:
.\"
.\"   Free Software Foundation, Inc.
.\"   51 Franklin Street, Fifth Floor
.\"   Boston, MA  02110-1301, USA.
.\"
.ds g \" empty
.ds G \" empty
.de Tp
.ie \\n(.$=0:((0\\$1)*2u>(\\n(.1u-\\n(.iu)) .TP
.el .TP "\\$1"
..
.TH GAPUTIL 1 "08 Jul 2015" "gaputil version 20150708"
.SH NAME
gaputil \- build nonuniform gap-sampling schedules

.SH SYNOPSIS
.B gaputil
\fIdensity\fR \fIN1\fR [\fIN2\fR [\fIN3\fR]] \fIgapfunc\fR

.SH DESCRIPTION
.PP
Build a nonuniform sampling (NUS) schedule on a grid having one, two or
three dimensions. The global sampling density of the schedule must be
specified in \fIdensity\fR as a value between 0 and 1. The arguments
\fIN1\fR, \fIN2\fR and \fIN3\fR are used to specify the size(s) of the
sampling grid to build the schedule upon.
.PP
Finally, the \fIgapfunc\fR argument must be supplied as a string (text)
that holds the gap equation. It is recommended that the gap equation be
placed in single quotes in order to ensure proper parsing.

.SH "GAP EQUATIONS"
Gap equations are defined in the Julia programming language. At program
startup, \fBgaputil\fR hands the value specified in \fIgapfunc\fR to a
Julia interpreter, which compiles it for later evaluation. All gap
equations must obey the following interface:

.in +4n
.nf

g(x::Float64, d::Int32,
  O::Array{Float64,1},
  N::Array{Float64,1},
  L::Float64)
.fi
.in

where \fBx\fR is the current sequence term, \fBd\fR is the dimension along
which the gap sampling algorithm is placing points, \fBO\fR is an array of
origin values, \fBN\fR is an array of grid sizes, and \fBL\fR is a scaling
factor that the gap sampling algorithm will adjust in order to achieve the
desired global sampling density. The arrays \fBO\fR and \fBN\fR both contain
an element for each dimension in the sampled grid, and may be indexed with
\fIe.g.\fR \fBd\fR.

.PP
The simplest possible gap equation is a uniform lattice:
.in +4n
.nf

L
.fi
.in

.PP
For another example, the sine-gap equation is as follows:
.in +4n
.nf

L * sin((pi / 2) * (x + sum(O)) / sum(N))
.fi
.in

.PP
The sine-burst equation is a bit longer:
.in +4n
.nf

L * sin((pi / 2) * (x + sum(O)) / sum(N))
  * sin((pi / 4) * N[d] * (x + sum(O)) / sum(N))^2
.fi
.in

.PP
For simplicity, the \fBsinegap()\fR and \fBsineburst()\fR methods are
predefined within the gap utility. To use these functions, just specify
the \fIgapfunc\fR argument as \fBsinegap(x,d,O,N,L)\fR or
\fBsineburst(x,d,O,N,L)\fR when running the utility.

.PP
For more detailed information on the accepted syntax of gap equations,
consult the Julia language documentation.

.SH NOTE
The utility is currently soft-limited to building up to three-dimensional
schedules, but this may be trivially changed in the source code prior to
compilation. See \fIgaputil.h\fR for more information.

.SH CITING
If you use the gap utility in any published work, it is requested that you
provide attribution to the following peer-reviewed article:

.in +4n
.nf

\fBBradley Worley and Robert Powers\fR,
Deterministic Multidimensional Nonuniform Gap Sampling,
\fIJournal of Magnetic Resonance\fR, 2015, submitted.
.fi
.in

.SH AUTHOR
Bradley Worley <geekysuavo@gmail.com>

.SH COPYRIGHT
Copyright \(co 2015 Bradley Worley <geekysuavo@gmail.com>
.br
This is free software. You may redistribute copies of it under the terms of
version 2.0 of the GNU General Public License
<http://www.gnu.org/licenses/old-licenses/gpl-2.0.html>.
There is NO WARRANTY, to the extent permitted by law.

.SH "SEE ALSO"
.BR julia(1)


gaputil.c

/* gaputil: generalized gap sampling schedule generation utility.
 *
 * Copyright (C) 2015 Bradley Worley <geekysuavo@gmail.com>.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to:
 *
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301, USA.
 */

/* include the main header. */
#include "gaputil.h"

/* main(): application entry point.
 *
 * arguments:
 *  @argc: number of command line arguments.
 *  @argv: command line argument string array.
 *
 * returns:
 *  integer representing whether execution terminated without error (0)
 *  or not (1).
 */
int main (int argc, char **argv) {
  /* declare variables to hold schedule parameters:
   *  @D: total number of Nyquist grid dimensions.
   *  @N: tuple holding the Nyquist grid sizes.
   *  @d: effective sampling density, in (0,1).
   */
  unsigned int D;
  tuple_t N;
  double d;

  /* declare variables to hold schedule values:
   *  @xlst: tuple of linear indices in the schedule.
   *  @xt: tuple to hold unpacked linear indices.
   */
  tuple_t xlst, xt;

  /* declare a general-purpose loop index variable:
   *  @i: loop counter and iteration index.
   */
  unsigned int i;

  /* determine the number of grid dimensions. */
  D = argc - 3;

  /* check that a supported number of dimensions was requested.
   *
   * the GAPUTIL_DIMS_MIN and GAPUTIL_DIMS_MAX preprocessor definitions,
   * located in gaputil.h, are used to soft-limit the number of dimensions
   * that the program will build sampling schedules upon.
   */
  if (D < GAPUTIL_DIMS_MIN || D > GAPUTIL_DIMS_MAX) {
    /* output a usage statement and return failure. */
    fprintf(stderr, GAPUTIL_USAGE, argv[0]);
    return 1;
  }

  /* read in the sampling density. */
  d = atof(argv[1]);

  /* validate the sampling density. */
  if (d <= 0.0 || d >= 1.0) {
    /* output an error and return failure. */
    fprintf(stderr, "error: sampling density must lie in (0,1)\n");
    return 1;
  }

  /* allocate the grid size tuple. */
  if (!tupalloc(&N, D) || !tupalloc(&xt, D)) {
    /* output an error and return failure. */
    fprintf(stderr, "error: failed to allocate grid size tuple\n");
    return 1;
  }

  /* read in the grid sizes. */
  for (i = 0; i < tupsize(&N); i++) {
    /* read the currently indexed argument. */
    tupset(&N, i, atoi(argv[i + 2]));

    /* validate the grid size. */
    if (tupget(&N, i) == 0) {
      /* output an error and return failure. */
      fprintf(stderr, "error: invalid N%u grid size\n", i + 1);
      return 1;
    }
  }

  /* initialize the julia interpreter. */
  jl_init(JULIA_INIT_DIR);

  /* build the final schedule array. */
  if (!seq(argv[argc - 1], &N, d, &xlst)) {
    /* output an error and return failure. */
    fprintf(stderr, "error: failed to compute output sequence\n");
    return 1;
  }

  /* sort the final schedule array and remove duplicate entries. */
  tupuniq(&xlst);

  /* print the final schedule values. */
  for (i = 0; i < tupsize(&xlst); i++) {
    /* unpack and print the current schedule value. */
    tupunpack(tupget(&xlst, i), &N, &xt);
    tupprint(&xt, stdout);
  }

  /* free the allocated tuples. */
  tupfree(&xlst);
  tupfree(&xt);
  tupfree(&N);

  /* return successfully. */
  termfree();
  return 0;
}

